avail-server/src/main/kotlin/com/avail/server/io/files/ServerFileWrapper.kt

```diff
@@ -0,0 +1,409 @@
+	 * the [TracedAction] that is [undoStackDepth] + 1 from the top of the
+	 * stack.
+	 */
+	fun undo ()
+	{
+		if (tracedActionStack.size < undoStackDepth + 1)
+		{
+			// We are at the bottom of the stack; there are no TracedActions
+			// eligible to be reverted.
+			return
+		}
+		tracedActionStack[tracedActionStack.size + ++undoStackDepth].revert(file)
+	}
+
+	/**
+	 * If an [undo] resulted in a [revert][TracedAction.reverseAction] on the
+	 * [file] and no other [FileAction] has occurred since,
+	 * [redo][TracedAction.redo] the previously reverted action.
+	 */
+	// TODO make thread safe?
+	fun redo ()
+	{
+		if (undoStackDepth == 0)
+		{
+			// There have been no TracedActions undone.
+			return
+		}
+		val index = tracedActionStack.size - undoStackDepth
+		// TODO validate in range
+		tracedActionStack[index].redo(file)
+		undoStackDepth--
+	}
+
+	/**
+	 * Delete the [file] from disk.
+	 *
+	 * @param id
+	 *   The [FileManager] cache id for this file.
+	 * @param failure
+	 *   A function that accepts a [ServerErrorCode] that describes the nature
+	 *   of the failure and an optional [Throwable]. TODO refine error handling
+	 */
+	fun delete(id: UUID, failure: (ServerErrorCode, Throwable?) -> Unit)
+	{
+		isReady.set(false)
+		try
+		{
+			FileManager.remove(id)
+			if (file.isOpen)
+			{
+				try
+				{
+					file.close()
+				}
+				catch (e: IOException)
+				{
+					// Do nothing
+				}
+			}
+			if (!Files.deleteIfExists(Paths.get(path)))
+			{
+				failure(ServerErrorCode.FILE_NOT_FOUND, null)
+			}
+		}
+		catch (e: IOException)
+		{
+			failure(ServerErrorCode.IO_EXCEPTION, e)
+		}
+	}
+
+	init
+	{
+		FileManager.executeFileTask (Runnable {
+			val p = Paths.get(path)
+			val mimeType = Tika().detect(p)
+			file = createFile(path, fileChannel, mimeType, this)
+		})
+	}
+
+	/** Close this [ServerFileWrapper]. */
+	fun close () { file.close() }
+
+	/**
+	 * Provide the [raw bytes][AvailServerFile.rawContent] of the enclosed
+	 * [AvailServerFile] to the requesting consumer.
+	 *
+	 * @param id
+	 *   The [FileManager.fileCache] [UUID] that uniquely identifies the file.
+	 * @param consumer
+	 *   A function that accepts the [FileManager.fileCache] [UUID] that
+	 *   uniquely identifies the file and the
+	 *   [raw bytes][AvailServerFile.rawContent] of an [AvailServerFile].
+	 * @param failureHandler
+	 *   A function that accepts a [ServerErrorCode] that describes the nature
+	 *   of the failure and an optional [Throwable]. TODO refine error handling
+	 */
+	fun provide(
+		id: UUID,
+		consumer: (UUID, ByteArray) -> Unit,
+		failureHandler: (ServerErrorCode, Throwable?) -> Unit)
+	{
+		// TODO should be run on a separate thread or is that handled inside the
+		//  consumer?
+		if (!isLoadingFile)
+		{
+			// Opportunistic
+			consumer(id, file.rawContent)
+			return
+		}
+		synchronized(this)
+		{
+			if (!isLoadingFile)
+			{
+				// Opportunistic
+				consumer(id, file.rawContent)
+				return
+			}
```

Comment:
The double-checked locking optimization is kind of a dangerous pattern, and this is an incorrect implementation anyway — it's not safe to invoke `consumer` outside of the lock. The "correct" implementation of the pattern is:

```kt
if (!isLoadingFile)
{
	synchronized(this)
	{
		is (!isLoadingFile)
		{
			consumer(id, file.rawContent)
			return
		}
	}
}
```

Note that the only part of this that might be safe to repeat is the actual predicate test, definitely not the protected body. It's probably safest to just unconditionally acquire the lock, unless there is reason to believe that this operation should be highly contentious among competing threads (which it shouldn't be).

Reply:
fixed